Merge two config maps into one that defines control switches and user defined injections

MichalGuzieniuk opened the pull request:
This pull requests aims to merge User Defined Injections (UDI) ConfigMap into the control switches ConfigMap.

In result, there will be only one map to provided by the user. From the code perspective I refactored code by moving out from webhook.go all functions related to UDI into separate package called custominjections. This way, there is a clear information about package responsibility.

Fixes https://github.com/k8snetworkplumbingwg/network-resources-injector/issues/61

Eoghan1232 commented:
@zshi-redhat Could you review?

zshi-redhat commented on pkg/custominjections/custominjections.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// SetCustomizedInjections sets additional injections to be applied in Pod spec
+func (userDefinedInjects *UserDefinedInjections) SetCustomizedInjections(injectionsCm *corev1.ConfigMap) {
```
Please use the consistent interface name as in line 95.
Either change `userDefinedInjects` to `injections` in this line or change `injections` to `userDefinedInjects` in line 95.

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/custominjections/custominjections.go:
```diff
@@ -0,0 +1,112 @@
+package custominjections
```
I noticed there are some remainings when I updated the `custom injection` term to `user defined injection`. Would you mind updating this PR to use `user defined injection`? Ideally we would like to replace custom to user-defined.

MichalGuzieniuk replied:
Done.

zshi-redhat reviewed (commented):
@MichalGuzieniuk Thanks for the PR and sorry for the late review!

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// SetCustomizedInjections sets additional injections to be applied in Pod spec
+func (userDefinedInjects *UserDefinedInjections) SetCustomizedInjections(injectionsCm *corev1.ConfigMap) {
```
SetCustomizedInjections -> SetUserDefinedInjections

MichalGuzieniuk replied:
Done.

I should do a training from search tool ;) - how could I missed that. Sorry for double work on review.

MichalGuzieniuk replied:
Done

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// CreateCustomizedPatch creates customized patch for the specified POD
+func (userDefinedInjects *UserDefinedInjections) CreateCustomizedPatch(pod corev1.Pod) ([]types.JsonPatchOperation, error) {
```
CreateCustomizedPatch -> CreateUserDefinedPatch

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+	. "github.com/onsi/gomega"
+)
+
+func TestCustomInjections(t *testing.T) {
```
TestUserDefinedInjections

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+
+func TestCustomInjections(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "CustomInjections Suite")
```
CustomInjections -> UserDefinedInjections

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections_test.go:
```diff
@@ -0,0 +1,246 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("CustomInjections", func() {
```
CustomInjections > UserDefinedInjections

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections_test.go:
```diff
@@ -0,0 +1,246 @@
+		func(pod corev1.Pod, userDefinedInjectPatchs map[string]types.JsonPatchOperation, out []types.JsonPatchOperation) {
+			userDefinedInjects := CreateUserInjectionsStructure()
+			userDefinedInjects.Patchs = userDefinedInjectPatchs
+			appliedPatchs, _ := userDefinedInjects.CreateCustomizedPatch(pod)
```
CreateCustomizedPatch > CreateUserDefinedPatch

zshi-redhat commented on pkg/userdefinedinjections/userdefinedinjections_test.go:
```diff
@@ -0,0 +1,246 @@
+		func(in *corev1.ConfigMap, existing map[string]types.JsonPatchOperation, out map[string]types.JsonPatchOperation) {
+			userDefinedInjects := CreateUserInjectionsStructure()
+			userDefinedInjects.Patchs = existing
+			userDefinedInjects.SetCustomizedInjections(in)
```
SetCustomizedInjections -> SetUserDefinedInjections

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/webhook/webhook.go:
```diff
@@ -735,12 +709,12 @@ func appendAddAnnotPatch(patch []jsonPatchOperation, pod corev1.Pod, userDefined
 }
 
-func appendCustomizedPatch(patch []jsonPatchOperation, pod corev1.Pod, userDefinedPatch []jsonPatchOperation) []jsonPatchOperation {
+func appendCustomizedPatch(patch []types.JsonPatchOperation, pod corev1.Pod, userDefinedPatch []types.JsonPatchOperation) []types.JsonPatchOperation {
```
appendCustomizedPatch > appendUserDefinedPatch

MichalGuzieniuk replied:
Done.

zshi-redhat commented on pkg/webhook/webhook.go:
```diff
@@ -788,7 +762,7 @@ func MutateHandler(w http.ResponseWriter, req *http.Request) {
 	glog.Infof("AdmissionReview request received for pod %s/%s", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name)
 
-	userDefinedPatch, err := createCustomizedPatch(pod)
+	userDefinedPatch, err := userDefinedInjections.CreateCustomizedPatch(pod)
```
CreateCustomizedPatch > CreateUserDefinedPatch

MichalGuzieniuk replied:
Done.

zshi-redhat commented on cmd/webhook/main.go:
```diff
@@ -207,16 +210,7 @@ func main() {
-				}
-			}
-			webhook.SetCustomizedInjections(cm)
+			userInjections.SetCustomizedInjections(cm)
```
SetCustomizedInjections -> SetUserDefinedInjections

MichalGuzieniuk replied:
Done.

zshi-redhat commented on README.md:
```diff
@@ -263,10 +263,21 @@ Below is an example of user defined injection ConfigMap:
-  feature.pod.kubernetes.io_sriov-network: '{"op": "add", "path": "/metadata/annotations", "value": {"k8s.v1.cni.cncf.io/networks": "sriov-net-attach-def"}}'
+  config.json: |
+    {
+      "customInjections": {
```
customInjections > user-defined-injections

MichalGuzieniuk replied:
Done.

zshi-redhat reviewed (commented):
A few more places to replace custom with user-defined. see comments.

zshi-redhat commented:
In the `cmd/webhook/main.go`, there is this piece of code:

```
202                 case <-time.After(30 * time.Second):
203                         cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
204                                 context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
205                         if err != nil {
206                                 glog.Warningf("Error within control switches map %s", err.Error())
207                                 if !errors.IsNotFound(err) {
208                                         glog.Info("Map with runtime configuration not available using default / CLI settings")
209                                 }
210                         }
211                         // has to be called each time because we need to restore default config when map is empty
212                         controlSwitches.ProcessControlSwitchesConfigMap(cm)
213                         userInjections.SetUserDefinedInjections(cm)
214                 }
```

Could we use the following:

```
                        cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
                                context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
                        if err != nil {
                                if !errors.IsNotFound(err) {
                                        glog.Warningf("Error getting control switches configmap %s", err.Error())
                                }
                                continue
                        }   
                        // has to be called each time because we need to restore default config when map is empty
                        controlSwitches.ProcessControlSwitchesConfigMap(cm)
                        userInjections.SetUserDefinedInjections(cm)
                }
```

So that it won't print a Warningf message every 30 seconds and the processing of controlSwitch configmap would be skipped, when there is no controlSwitch configmap defined.

Also could you merge the three commits into one?

Thank you!

MichalGuzieniuk commented:
> In the `cmd/webhook/main.go`, there is this piece of code:
> 
> ```
> 202                 case <-time.After(30 * time.Second):
> 203                         cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
> 204                                 context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
> 205                         if err != nil {
> 206                                 glog.Warningf("Error within control switches map %s", err.Error())
> 207                                 if !errors.IsNotFound(err) {
> 208                                         glog.Info("Map with runtime configuration not available using default / CLI settings")
> 209                                 }
> 210                         }
> 211                         // has to be called each time because we need to restore default config when map is empty
> 212                         controlSwitches.ProcessControlSwitchesConfigMap(cm)
> 213                         userInjections.SetUserDefinedInjections(cm)
> 214                 }
> ```
> 
> Could we use the following:
> 
> ```
>                 case <-time.After(30 * time.Second):
>                         cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
>                                 context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
>                         if err != nil {
>                                 if !errors.IsNotFound(err) {
>                                         glog.Warningf("Error getting control switches configmap %s", err.Error())
>                                 }
>                                 continue
>                         }   
>                         // has to be called each time because we need to restore default config when map is empty
>                         controlSwitches.ProcessControlSwitchesConfigMap(cm)
>                         userInjections.SetUserDefinedInjections(cm)
>                 }
> ```
> 
> So that it won't print a Warningf message every 30 seconds and the processing of controlSwitch configmap would be skipped, when there is no controlSwitch configmap defined.
> 
> Also could you merge the three commits into one?
> 
> Thank you!

I have removed the log that was polluting the events log. But please notice that I did not add **continue**'in case where there  error. According to what we agree within issue description, in case when there is no config map or it is invalid - NRI should restore default values.

I squashed also all commits into one, but could you explain me why I should do it? I thought that it is possible to squash commits when merge is done from my fork to npwg repo directly by gihub.

zshi-redhat commented:
> I have removed the log that was polluting the events log. But please notice that I did not add **continue**'in case where there error. According to what we agree within issue description, in case when there is no config map or it is invalid - NRI should restore default values.

In case there is no configmap - yes, I think it should re-initialize the settings.
In case the configmap is invalid - yes, NRI should restore default values (how do we define invalid, does it mean the data of configmap is not in right format, but configmap itself is retrievable by NRI?).

I think there is a third condition:
In case the NRI fails to get the configmap due to a temporary api unaccessibility, the settings will be re-initialized until the next api call success. I assume we would want to avoid the temporary change on the settings.

> 
> I squashed also all commits into one, but could you explain me why I should do it? I thought that it is possible to squash commits when merge is done from my fork to npwg repo directly by gihub.

Yes, you're right, the squash can be done in merge action, I will do that next time.

MichalGuzieniuk commented:
> In case there is no configmap - yes, I think it should re-initialize the settings. In case the configmap is invalid - yes, NRI should restore default values (how do we define invalid, does it mean the data of configmap is not in right format, but configmap itself is retrievable by NRI?).
> 
> I think there is a third condition: In case the NRI fails to get the configmap due to a temporary api unaccessibility, the settings will be re-initialized until the next api call success. I assume we would want to avoid the temporary change on the settings.
> 


Done.

zshi-redhat commented:
@Eoghan1232 would you want to take another look since there has been some update since your last approve?

Eoghan1232 commented:
> @Eoghan1232 would you want to take another look since there has been some update since your last approve?

Yes, I will review once more.
